src/postings/block_segment_postings.rs

```diff
@@ -24,7 +24,7 @@ fn max_score<I: Iterator<Item = Score>>(mut it: I) -> Option<Score> {
 #[derive(Clone)]
 pub struct BlockSegmentPostings {
```

Comment:
the `skip_reader` member in `BlockSegmentPostings` is `pub(crate)`, but calling next on it can break the invariant you rephrased ("if `block_loaded` is set to true the block loaded in `doc_decoder` and `freq_decoder` is matching the block of skip_reader).

Can you see if it is possible to make it private, and if needed expose it with an inline non mut accessor.